### PR TITLE
SONARJAVA-3769: Remove record fields from reporting in S3011(Reflection fields update)

### PR DESCRIPTION
**Pull request by dorian-burihabwa-sonarsource**

### CODE REVIEW

**Review by sebastian-hungerecker-sonarsource — CHANGES_REQUESTED** [review]
This looks good to me. I've made some suggestions for improvements and, as discussed, I won't merge until after the companion rule has been implemented because the excluded cases are still useful until the other rule is implemented to cover them.

You should also add test cases to bring up the coverage.

File: java-checks/src/main/java/org/sonar/java/checks/AccessibilityChangeCheck.java [hunk 1]
```diff
@@ -44,20 +51,109 @@
...
+
   @Override
   protected MethodMatchers getMethodInvocationMatchers() {
     return METHOD_MATCHERS;
   }
 
   @Override
   protected void onMethodInvocationFound(MethodInvocationTree mit) {
-    if(mit.symbol().name().equals("setAccessible")) {
+    if (isJava16OrGreater && isModifyingFieldFromRecord(mit)) {
```
[the remark this answers is not in the document]
Comment (dorian-burihabwa-sonarsource) [reply java-checks/src/main/java/org/sonar/java/checks/AccessibilityChangeCheck.java:80]: That sounds fair, the version check will be removed.

Please note that It looks like the Java version chosen in our tests does not enable the preview features. For instance, the [local type inference with var](https://github.com/SonarSource/sonar-java/blob/a1be09304b53d67f6c45182a9f4f83302fa58e80/java-checks/src/test/java/org/sonar/java/checks/VarCanBeUsedCheckTest.java#L37) needs to set the version to 10 for the `var` keyword to be properly processed.

File: java-checks/src/main/java/org/sonar/java/checks/AccessibilityChangeCheck.java [hunk 2]
```diff
@@ -44,20 +51,109 @@
...
       reportIssue(mit, "This accessibility bypass should be removed.");
     }
   }
 
+  private static boolean isModifyingFieldFromRecord(MethodInvocationTree mit) {
+    if (!mit.symbol().owner().type().is(JAVA_LANG_REFLECT_FIELD)) {
+      return false;
+    }
+    ExpressionTree fieldModificationExpression = mit.methodSelect();
+    MemberSelectExpressionTree memberSelect = (MemberSelectExpressionTree) fieldModificationExpression;
```
[the remark this answers is not in the document]
Comment (dorian-burihabwa-sonarsource) [reply java-checks/src/main/java/org/sonar/java/checks/AccessibilityChangeCheck.java:95]: Will fix that

**Comment by dorian-burihabwa-sonarsource** [comment]
These changes have been made part of SONARJAVA-3768 (#3673).

**Pull request closed without merging**